test/test_io.py

```diff
@@ -64,234 +64,290 @@ def temp_video(num_frames, height, width, fps, lossless=False, video_codec=None,
-            video, audio, info = io.read_video(f.name, pts_unit='sec')
-            # check that size is not equal to 5, but 3
-            # TODO fix this
-            if get_video_backend() == 'pyav':
-                self.assertEqual(len(video), 3)
-            else:
-                self.assertEqual(len(video), 4)
-            # but the valid decoded content is still correct
-            assert_equal(video[:3], data[:3])
-            # and the last few frames are wrong
-            with self.assertRaises(AssertionError):
-                assert_equal(video, data)
-
-    @unittest.skipIf(sys.platform == 'win32', 'temporarily disabled on Windows')
-    def test_write_video_with_audio(self):
-        f_name = os.path.join(VIDEO_DIR, "R6llTwEh07w.mp4")
-        video_tensor, audio_tensor, info = io.read_video(f_name, pts_unit="sec")
-
-        with get_tmp_dir() as tmpdir:
-            out_f_name = os.path.join(tmpdir, "testing.mp4")
-            io.video.write_video(
-                out_f_name,
-                video_tensor,
-                round(info["video_fps"]),
-                video_codec="libx264rgb",
-                options={'crf': '0'},
-                audio_array=audio_tensor,
-                audio_fps=info["audio_fps"],
-                audio_codec="aac",
-            )
-
-            out_video_tensor, out_audio_tensor, out_info = io.read_video(
-                out_f_name, pts_unit="sec"
-            )
-
-            self.assertEqual(info["video_fps"], out_info["video_fps"])
-            assert_equal(video_tensor, out_video_tensor)
-
-            audio_stream = av.open(f_name).streams.audio[0]
-            out_audio_stream = av.open(out_f_name).streams.audio[0]
-
-            self.assertEqual(info["audio_fps"], out_info["audio_fps"])
-            self.assertEqual(audio_stream.rate, out_audio_stream.rate)
-            self.assertAlmostEqual(audio_stream.frames, out_audio_stream.frames, delta=1)
-            self.assertEqual(audio_stream.frame_size, out_audio_stream.frame_size)
-
-    # TODO add tests for audio
+            assert len(lv) == 4
+            assert_equal(data[4:8], lv)
+
+
+@pytest.mark.skipif(get_video_backend() != "pyav" and not io._HAS_VIDEO_OPT,
+                    reason="video_reader backend not available")
+@pytest.mark.skipif(av is None, reason="PyAV unavailable")
+def test_read_partial_video_bframes():
+    # do not use lossless encoding, to test the presence of B-frames
+    options = {'bframes': '16', 'keyint': '10', 'min-keyint': '4'}
+    with temp_video(100, 300, 300, 5, options=options) as (f_name, data):
+        pts, _ = io.read_video_timestamps(f_name)
+        for start in range(0, 80, 20):
+            for offset in range(1, 4):
+                lv, _, _ = io.read_video(f_name, pts[start], pts[start + offset - 1])
+                s_data = data[start:(start + offset)]
+                assert len(lv) == offset
+                assert_equal(s_data, lv, rtol=0.0, atol=TOLERANCE)
```

Comment:
Hmm, from my stand point `assert_equal` should only be used without tolerances (or better with the default `rtol = atol = 0.0`), because otherwise we wouldn't compare for equality.